tracing_test.go

```diff
@@ -21,314 +21,531 @@
+func (h *RawHandler) Handle(ctx context.Context, args *raw.Args) (*raw.Res, error) {
+	req := new(TracingRequest).fromRaw(args)
+	res, err := h.handleCall(ctx, req,
+		func(ctx context.Context, req *TracingRequest) *TracingResponse {
+			_, arg3, _, err := raw.Call(ctx, h.ch, h.ch.PeerInfo().HostPort,
+				h.ch.PeerInfo().ServiceName, "rawcall", nil, req.toRaw())
+			require.NoError(h.t, err)
+			return new(TracingResponse).fromRaw(h.t, arg3)
+		})
+	require.NoError(h.t, err)
+	return res.toRaw(h.t), nil
+}
 
-		ctx, cancel := json.NewContext(time.Second)
-		defer cancel()
+func (h *RawHandler) OnError(ctx context.Context, err error) { h.t.Errorf("onError %v", err) }
 
-		peer := ch.Peers().GetOrAdd(ch.PeerInfo().HostPort)
-		var response TracingResponse
-		require.NoError(t, json.CallPeer(ctx, peer, ch.PeerInfo().ServiceName, "call", &TracingRequest{
-			ForwardCount: 1,
-		}, &response))
-
-		clientSpan := CurrentSpan(ctx)
-		require.NotNil(t, clientSpan)
-		assert.Equal(t, uint64(0), clientSpan.ParentID())
-		assert.NotEqual(t, uint64(0), clientSpan.TraceID())
-		assert.True(t, clientSpan.TracingEnabled(), "Tracing should be enabled")
-		assert.Equal(t, clientSpan.TraceID(), response.TraceID)
-		assert.Equal(t, clientSpan.SpanID(), response.ParentID)
-		assert.True(t, response.TracingEnabled, "Tracing should be enabled")
-		assert.Equal(t, response.TraceID, response.SpanID, "traceID = spanID for root span")
-
-		nestedResponse := response.Child
-		require.NotNil(t, nestedResponse)
-		assert.Equal(t, clientSpan.TraceID(), nestedResponse.TraceID)
-		assert.Equal(t, response.SpanID, nestedResponse.ParentID)
-		assert.True(t, response.TracingEnabled, "Tracing should be enabled")
-		assert.NotEqual(t, response.SpanID, nestedResponse.SpanID)
-	})
+// JSONHandler tests tracing over JSON encoding
+type JSONHandler struct {
+	traceHandler
 }
 
-func TestTraceReportingEnabled(t *testing.T) {
-	initialTime := time.Date(2015, 2, 1, 10, 10, 0, 0, time.UTC)
+func (h *JSONHandler) callJSON(ctx json.Context, req *TracingRequest) (*TracingResponse, error) {
+	return h.handleCall(ctx, req, func(ctx context.Context, req *TracingRequest) *TracingResponse {
+		jctx := ctx.(json.Context)
+		sc := h.ch.Peers().GetOrAdd(h.ch.PeerInfo().HostPort)
+		childResp := new(TracingResponse)
+		require.NoError(h.t, json.CallPeer(jctx, sc, h.ch.PeerInfo().ServiceName, "call", req, childResp))
+		return childResp
+	})
+}
 
-	var state struct {
-		signal chan struct{}
+func (h *JSONHandler) onError(ctx context.Context, err error) { h.t.Errorf("onError %v", err) }
 
-		call TraceData
-		span Span
-	}
-	testTraceReporter := TraceReporterFunc(func(data TraceData) {
-		defer close(state.signal)
+// ThriftHandler tests tracing over Thrift encoding
+type ThriftHandler struct {
+	traceHandler
+}
 
-		span := data.Span
-		data.Span = Span{}
-		state.call = data
-		state.span = span
+func (h *ThriftHandler) Call(ctx thrift.Context, arg *gen.Data) (*gen.Data, error) {
```

Comment:
My preference is to have the tchannel tests not rely on the Thrift package. There's a couple of reasons:
- Adds another dependency to the tchannel tests, which will make running `go test .` in the core tchannel package a little slower
- Package boundaries -- there may be a proto implementation that lives outside of the package, and they should be able to do what Thrift does and have similar tests without having to touch the tchannel package

My preference is:
 - Unit test the functions exposed in the tchannel package using fake spans and calling them directly
 - Move the thrift/json integration tests into their respective packages which will ensure they're hooked up correctly
 - Feel free to add `testutils/tracetest` or some package with common code if they share a lot of logic.